// @remixproject/plugin-ws/lib/ws.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createClient = exports.WebsocketConnector = void 0;
const plugin_1 = require("@remixproject/plugin");
/**
 * This Websocket connector works with the library `ws`
 */
class WebsocketConnector {
    constructor(websocket) {
        this.websocket = websocket;
    }
    /** Send a message to the engine */
    send(message) {
        this.websocket.send(JSON.stringify(message));
    }
    /** Get messae from the engine */
    on(cb) {
        this.websocket.on('message', (event) => cb(JSON.parse(event)));
    }
}
exports.WebsocketConnector = WebsocketConnector;
/**
 * Connect a Websocket plugin client to a web engine
 * @param client An optional websocket plugin client to connect to the engine.
 *
 * ---------
 * @example
 * ```typescript
 * const wss = new WebSocket.Server({ port: 8080 });
 * wss.on('connection', (ws) => {
 *  const client = createClient(ws)
 * })
 * ```
 * ---------
 * @example
 * ```typescript
 * class MyPlugin extends PluginClient {
 *  methods = ['hello']
 *  hello() {
 *   console.log('Hello World')
 *  }
 * }
 * const wss = new WebSocket.Server({ port: 8080 });
 * wss.on('connection', (ws) => {
 *  const client = createClient(ws, new MyPlugin())
 * })
 * ```
 */
exports.createClient = (websocket, client = new plugin_1.PluginClient()) => {
    const c = client;
    plugin_1.connectClient(new WebsocketConnector(websocket), c);
    plugin_1.applyApi(c);
    return c;
};
